File: php/src/Caller.php
<?php

declare(strict_types=1);

namespace Unilog;

/**
 * Short caller trace — attached to every regular log call (not the full
 * exception.stacktrace, which stays reserved for actual exceptions).
 * code.function/code.filepath/code.lineno for the immediate call site, plus
 * code.stacktrace — up to LOG_CALLER_TRACE_FRAMES (default 3) frames above
 * it, closest-first, skipping this package's own frames.
 */
final class Caller
{
    public static function enabled(): bool
    {
        return Env::flag('LOG_CALLER_INFO');
    }

    private static function frameLimit(): int
    {
        return Env::int('LOG_CALLER_TRACE_FRAMES', 3);
    }

    private static function isSkippedWrapper(string $file): bool
    {
        foreach (self::SKIP_SUBSTRINGS as $needle) {
            if (str_contains($file, $needle)) {
                return true;
            }
        }
        return false;
    }

    private const PACKAGE_DIR = __DIR__;

    // Well-known logging wrapper libraries whose own frames sit between the
    // user's real call site and here (psr/log's LoggerTrait delegates every
    // named level method through log(); Monolog\Logger::addRecord walks its
    // own handler stack before reaching UnilogHandler::write()) — skipped by
    // path substring so the "nearest" frame is the user's code, not a
    // wrapper's. Harmless if a project doesn't have them installed at all.
    private const SKIP_SUBSTRINGS = ['/psr/log/src/', '/monolog/monolog/src/'];

    /**
     * @return array<string, mixed>
     */
    public static function attributes(): array
    {
        $limit = self::frameLimit();
        // DEBUG_BACKTRACE_IGNORE_ARGS: argument values can contain secrets
        // (passwords, tokens) — never capture them just to compute a caller trace.
        //
        // Filtered by each frame's FILE (the location the call was made
        // FROM), not by 'function'/'class' — PHP's backtrace pairs a frame's
        // function name with the NEXT frame's file/line (frame N's function
        // is "what's currently running", frame N's file/line is "where IT
        // was called from"), which makes a name-based filter miss exactly
        // the boundary frame where Unilog's own code calls into user code.
        // File-based filtering has no such off-by-one.
        $trace = debug_backtrace(\DEBUG_BACKTRACE_IGNORE_ARGS, $limit + 16);

        $frames = [];
        foreach ($trace as $frame) {
            if (!isset($frame['file']) || str_starts_with($frame['file'], self::PACKAGE_DIR)) {
                continue;
            }
            if (self::isSkippedWrapper($frame['file'])) {
                continue;
            }
            $fn = isset($frame['class']) ? $frame['class'] . $frame['type'] . $frame['function'] : ($frame['function'] ?? '?');
            $frames[] = ['fn' => $fn, 'file' => $frame['file'], 'line' => $frame['line'] ?? 0];
            if (count($frames) >= $limit) {
                break;
            }
        }
        if ($frames === []) {
            return [];
        }
        $nearest = $frames[0];
        $attrs = [
            'code.function' => $nearest['fn'],
            'code.filepath' => $nearest['file'],
            'code.lineno' => $nearest['line'],
        ];
        if (count($frames) > 1) {
            $attrs['code.stacktrace'] = implode(' < ', array_map(
                static fn ($f) => sprintf('%s (%s:%d)', $f['fn'], basename($f['file']), $f['line']),
                $frames
            ));
        }
        return $attrs;
    }
}

File: python/unilog/_caller.py
"""Short caller trace — attached to every regular log call (not the full
`exception.stacktrace`, which stays reserved for actual exceptions). Mirrors
what the sanstv TSV format carried on every line (module::funcName,
filename:lineno) but as a bounded, OTel-attribute-shaped addition:
`code.function`/`code.filepath`/`code.lineno` for the immediate call site,
plus `code.stacktrace` — up to LOG_CALLER_TRACE_FRAMES (default 3) frames
above it, closest-first, skipping stdlib `logging` and unilog's own frames."""

from __future__ import annotations

import logging
import os
import traceback

_LOGGING_DIR = os.path.dirname(os.path.abspath(logging.__file__))
_PKG_DIR = os.path.dirname(os.path.abspath(__file__))


def _is_internal(filename: str) -> bool:
    filename = os.path.abspath(filename)
    return filename.startswith(_LOGGING_DIR) or filename.startswith(_PKG_DIR)


def caller_attributes(limit: int | None = None) -> dict[str, object]:
    """Walk the real call stack (not the LogRecord's own findCaller, which
    only gives one frame) and return code.function/filepath/lineno for the
    nearest application frame, plus a short code.stacktrace chain."""
    if limit is None:
        limit = int(os.environ.get("LOG_CALLER_TRACE_FRAMES", "3"))
    frames = [f for f in traceback.extract_stack()[:-1] if not _is_internal(f.filename)]
    if not frames:
        return {}
    app_frames = frames[-limit:]
    nearest = app_frames[-1]
    attrs: dict[str, object] = {
        "code.function": nearest.name,
        "code.filepath": nearest.filename,
        "code.lineno": nearest.lineno,
    }
    if len(app_frames) > 1:
        chain = " < ".join(f"{f.name} ({os.path.basename(f.filename)}:{f.lineno})" for f in reversed(app_frames))
        attrs["code.stacktrace"] = chain
    return attrs


def enabled() -> bool:
    return os.environ.get("LOG_CALLER_INFO", "1") != "0"
